File: generate-pdf.py
#!/usr/bin/env python3

"""
Script para gerar PDFs dos arquivos HTML usando Chrome headless

Instalação:
pip install selenium webdriver-manager

Uso:
python3 generate-pdf.py
"""

import os
import time
from selenium import webdriver
from selenium.webdriver.chrome.service import Service
from selenium.webdriver.chrome.options import Options
from webdriver_manager.chrome import ChromeDriverManager

def generate_pdf(html_file, output_pdf):
    """Gera PDF a partir de um arquivo HTML"""
    
    print(f"📄 Gerando PDF: {output_pdf}...")
    
    # Configura Chrome headless
    chrome_options = Options()
    chrome_options.add_argument('--headless=new')
    chrome_options.add_argument('--no-sandbox')
    chrome_options.add_argument('--disable-dev-shm-usage')
    chrome_options.add_argument('--disable-gpu')
    chrome_options.add_argument('--window-size=1400,10000')
    
    # Configurações de impressão para PDF
    settings = {
        "recentDestinations": [{
            "id": "Save as PDF",
            "origin": "local",
            "account": ""
        }],
        "selectedDestinationId": "Save as PDF",
        "version": 2,
        "isHeaderFooterEnabled": False,
        "isLandscapeEnabled": False,
        "isCssBackgroundEnabled": True,
        "mediaSize": {
            "height_microns": 297000,
            "width_microns": 210000
        }
    }
    
    prefs = {
        'printing.print_preview_sticky_settings.appState': settings,
        'savefile.default_directory': os.getcwd()
    }
    chrome_options.add_experimental_option('prefs', prefs)
    
    # Inicia Chrome
    service = Service(ChromeDriverManager().install())
    driver = webdriver.Chrome(service=service, options=chrome_options)
    
    try:
        # Carrega o arquivo HTML
        html_path = os.path.abspath(html_file)
        driver.get(f'file://{html_path}')
        
        # Aguarda 3 segundos para gráficos carregarem
        time.sleep(3)
        
        # Executa comando de impressão
        result = driver.execute_cdp_cmd('Page.printToPDF', {
            'printBackground': True,
            'landscape': False,
            'paperWidth': 8.27,  # A4 width in inches
            'paperHeight': 11.69,  # A4 height in inches
            'marginTop': 0.2,
            'marginBottom': 0.2,
            'marginLeft': 0.2,
            'marginRight': 0.2,
            'preferCSSPageSize': True
        })
        
        # Salva o PDF
        import base64
        with open(output_pdf, 'wb') as f:
            f.write(base64.b64decode(result['data']))
        
        print(f"✅ PDF gerado: {output_pdf}")
        
    finally:
        driver.quit()

def main():
    """Função principal"""
    
    print('🚀 Iniciando geração de PDFs...\n')
    
    files = [
        {
            'input': 'DESCRITIVO_SISTEMA_EXECUTIVO_CLEVEL.html',
            'output': 'DESCRITIVO_SISTEMA_EXECUTIVO_CLEVEL.pdf'
        },
        {
            'input': 'VARIAVEIS_AMBIENTE_TOKENS_MODERNO.html',
            'output': 'VARIAVEIS_AMBIENTE_TOKENS_MODERNO.pdf'
        }
    ]
    
    for file_info in files:
        input_file = file_info['input']
        output_file = file_info['output']
        
        if not os.path.exists(input_file):
            print(f"❌ Arquivo não encontrado: {input_file}")
            continue
        
        try:
            generate_pdf(input_file, output_file)
            print()
        except Exception as e:
            print(f"❌ Erro ao processar {input_file}: {str(e)}")
            print()
    
    print('✨ Processo concluído!')

if __name__ == '__main__':
    main()

File: generate-pdf.js
#!/usr/bin/env node

/**
 * Script para gerar PDFs e PNGs dos arquivos HTML
 * 
 * Instalação:
 * npm install -g puppeteer
 * 
 * Uso:
 * node generate-pdf.js
 */

const puppeteer = require('puppeteer');
const path = require('path');
const fs = require('fs');

const files = [
    {
        input: 'DESCRITIVO_SISTEMA_EXECUTIVO_CLEVEL.html',
        outputPdf: 'DESCRITIVO_SISTEMA_EXECUTIVO_CLEVEL.pdf',
        outputPng: 'DESCRITIVO_SISTEMA_EXECUTIVO_CLEVEL.png'
    },
    {
        input: 'VARIAVEIS_AMBIENTE_TOKENS_MODERNO.html',
        outputPdf: 'VARIAVEIS_AMBIENTE_TOKENS_MODERNO.pdf',
        outputPng: 'VARIAVEIS_AMBIENTE_TOKENS_MODERNO.png'
    }
];

async function generatePDF(inputFile, outputPdf) {
    console.log(`📄 Gerando PDF: ${outputPdf}...`);
    
    const browser = await puppeteer.launch({
        headless: 'new',
        args: ['--no-sandbox', '--disable-setuid-sandbox']
    });
    
    const page = await browser.newPage();
    
    // Carrega o arquivo HTML
    const htmlPath = path.resolve(__dirname, inputFile);
    await page.goto(`file://${htmlPath}`, {
        waitUntil: 'networkidle0'
    });
    
    // Aguarda os gráficos Chart.js carregarem
    await page.waitForTimeout(2000);
    
    // Gera o PDF
    await page.pdf({
        path: outputPdf,
        format: 'A4',
        printBackground: true,
        margin: {
            top: '0.5cm',
            right: '0.5cm',
            bottom: '0.5cm',
            left: '0.5cm'
        }
    });
    
    await browser.close();
    console.log(`✅ PDF gerado: ${outputPdf}`);
}

async function generatePNG(inputFile, outputPng) {
    console.log(`🖼️  Gerando PNG: ${outputPng}...`);
    
    const browser = await puppeteer.launch({
        headless: 'new',
        args: ['--no-sandbox', '--disable-setuid-sandbox']
    });
    
    const page = await browser.newPage();
    
    // Define viewport para captura de tela
    await page.setViewport({
        width: 1400,
        height: 10000,
        deviceScaleFactor: 2 // Retina quality
    });
    
    // Carrega o arquivo HTML
    const htmlPath = path.resolve(__dirname, inputFile);
    await page.goto(`file://${htmlPath}`, {
        waitUntil: 'networkidle0'
    });
    
    // Aguarda os gráficos Chart.js carregarem
    await page.waitForTimeout(2000);
    
    // Captura a altura total da página
    const bodyHandle = await page.$('body');
    const { height } = await bodyHandle.boundingBox();
    await bodyHandle.dispose();
    
    // Ajusta viewport para altura completa
    await page.setViewport({
        width: 1400,
        height: Math.ceil(height),
        deviceScaleFactor: 2
    });
    
    // Gera o PNG
    await page.screenshot({
        path: outputPng,
        fullPage: true
    });
    
    await browser.close();
    console.log(`✅ PNG gerado: ${outputPng}`);
}

async function main() {
    console.log('🚀 Iniciando geração de PDFs e PNGs...\n');
    
    for (const file of files) {
        const inputPath = path.resolve(__dirname, file.input);
        
        if (!fs.existsSync(inputPath)) {
            console.error(`❌ Arquivo não encontrado: ${file.input}`);
            continue;
        }
        
        try {
            await generatePDF(file.input, file.outputPdf);
            await generatePNG(file.input, file.outputPng);
            console.log('');
        } catch (error) {
            console.error(`❌ Erro ao processar ${file.input}:`, error.message);
        }
    }
    
    console.log('✨ Processo concluído!');
}

main().catch(console.error);

